WalletWasabi.Fluent/ViewModels/Wallets/Send/TransactionPreviewViewModel.cs

```diff
@@ -165,38 +183,51 @@ private async Task OnExportPsbtAsync()
 		}
 	}
 
-	private void UpdateTransaction(TransactionSummaryViewModel summary, BuildTransactionResult transaction)
+	private void UpdateTransaction(TransactionSummaryViewModel summary, BuildTransactionResult transaction, bool addToUndoHistory = true)
 	{
 		if (!summary.IsPreview)
 		{
+			if (addToUndoHistory && _transaction is { })
+			{
+				AddToUndoHistory();
+			}
+
 			_transaction = transaction;
 		}
 
-		summary.UpdateTransaction(transaction);
+		summary.UpdateTransaction(transaction, _info);
 
 		DisplayedTransactionSummary = summary;
 	}
 
 	private async Task OnAdjustFeeAsync()
 	{
+		AddToUndoHistory();
+
 		var feeRateDialogResult = await NavigateDialogAsync(new SendFeeViewModel(_wallet, _info, false));
 
 		if (feeRateDialogResult.Kind == DialogResultKind.Normal && feeRateDialogResult.Result is { } newFeeRate &&
 		    newFeeRate != _info.FeeRate)
 		{
 			_info.FeeRate = feeRateDialogResult.Result;
 
-			await BuildAndUpdateAsync(BuildTransactionReason.FeeChanged);
+			await BuildAndUpdateAsync(BuildTransactionReason.FeeChanged, false);
 		}
+		else
+		{
+			_undoHistory.Pop();
+		}
+
+		CanUndo = _undoHistory.Any();
```

Comment:
Can you elaborate on this? 

Why `AddToUndoHistory()` called here and after the `BuildAndUpdateAsync` is called with `addToUndoHistory = false` param?

Why not just simply call `BuildAndUpdateAsync`? That would call the `UpdateTransaction` where the `AddToUndoHistory` would be called.

And even dealing with the stack wouldn't be needed.

Reply:
@soosr `BuildAndUpdateAsync` needed to push the undo entry *before* the `_info.FeeRate = ` line, in order to store the current value. And `BuildAndUpdateAsync` needs to run *after* that.

Reply:
ahh okay, I understand the problem.

What do you think about the idea of... Every time a transaction gets built, we would store the current TransactionInfo's clone in a new variable `currentTransactionInfo`... so the original transaction info could be modified anytime and when you need to add a new undo item... you just use the `currentTransactionInfo` clone rather than making a clone at that point.

Storing `currentTransactionInfo` would fit inside of the `UpdateTransaction` where we store the new transaction `_transaction = transaction;`.

What do you think, can it work? It would simplify a lot I guess.

Reply:
@soosr 

> What do you think about the idea of... Every time a transaction gets built, we would store the current TransactionInfo's clone in a new variable currentTransactionInfo... so the original transaction info could be modified anytime and when you need to add a new undo item... you just use the currentTransactionInfo clone rather than making a clone at that point.

> Storing currentTransactionInfo would fit inside of the UpdateTransaction where we store the new transaction _transaction = transaction;.

Implemented this. Code got cleaner and simpler, but I can no longer see any change suggestions, so I couldn't test it thoroughly.

Reply:
> I can no longer see any change suggestions, so I couldn't test it thoroughly.

I have tested it, seems ok to me.